sed/io.py

```diff
@@ -179,3 +181,120 @@ def h5_to_xarray(faddr: str, mode: str = "r") -> xr.DataArray:
             xarray.attrs["metadata"] = metadata
 
         return xarray
+
+
+def array_to_tiff(array: np.ndarray, faddr: Union[Path, str]) -> None:
+    """Save array to imageJ compatible tiff stack.
+
+    Args:
+        array: array to save. 2D,3D or 4D
+        faddr: full path and name of file to save.
+
+    Raise:
+        NotImplementedError: if the input array is not 2,3 or 4 dimensional.
+    """
+    if array.ndim == 2:
+        out = np.expand_dims(array, [0, 1, 2, 5])
+    elif array.ndim == 3:
+        out = np.expand_dims(array, [0, 2, 5])
+    elif array.ndim == 4:
+        out = np.expand_dims(array, [2, 5])
+    else:
+        raise NotImplementedError("Only 2-3-4D arrays supported.")
+
+    tifffile.imwrite(faddr, out.astype(np.float32), imagej=True)
+    print(f"Successfully saved array with shape {array.shape} to {faddr}")
+
+
+def xarray_to_tiff(
+    data: xr.DataArray,
+    faddr: Union[Path, str],
+    axis_dict: dict = None,
+) -> None:
+    """Save an xarray as a  .tiff stack compatible with ImageJ
+
+    Args:
+        data: data to be saved. ImageJ likes tiff files with axis order as
+        TZCYXS. Therefore, best axis order in input should be: Time, Energy,
+        posY, posX. The channels 'C' and 'S' are automatically added and can
+        be ignored.
+        faddr: full path and name of file to save.
+        axis_dict: name pairs for correct axis ordering. Keys should be any of
+        T,Z,C,Y,X,S. The Corresponding value will be searched among the
+        dimensions of the xarray, and placed in the right order for imagej
+        stacks metadata.
+
+    Raise:
+        AttributeError: if more than one axis corresponds to a single dimension
+    """
+
+    assert isinstance(data, xr.DataArray), "Data must be an xarray.DataArray"
+    dims_to_add = {"C": 1, "S": 1}
+    dims_order = []
+
+    if axis_dict is None:
+        axis_dict = {
+            "T": ["delayStage", "pumpProbeTime", "time"],
+            "Z": ["dldTime", "energy"],
+            "C": ["C"],
+            "Y": ["dldPosY", "ky"],
+            "X": ["dldPosX", "kx"],
+            "S": ["S"],
+        }
+    else:
+        for key in ["T", "Z", "C", "Y", "X", "S"]:
+            if key not in axis_dict.keys():
+                axis_dict[key] = key
+
+    # Sort the dimensions in the correct order, and fill with one-point dimensions
+    # the missing axes.
+    for key in ["T", "Z", "C", "Y", "X", "S"]:
+        axis_name_list = [name for name in axis_dict[key] if name in data.dims]
+        if len(axis_name_list) > 1:
+            raise AttributeError(f"Too many dimensions for {key} axis.")
+        elif len(axis_name_list) == 1:
+            dims_order.append(*axis_name_list)
+        else:
+            dims_to_add[key] = 1
+            dims_order.append(key)
+
+    xres = data.expand_dims(dims_to_add)
+    xres = xres.transpose(*dims_order)
+    if ".tif" not in faddr:
+        faddr += ".tif"
```

Comment:
I would also allow ".tiff"

Reply:
Does this work if faddr is a Path?

Reply:
Also, maybe move this into the wrapper function.

Reply:
Actually it does not work when faddr is Path, and all because of that if... I think I'll just remove it as it is there to ensure the file name has the right extension. There is probably a better way to do it with Pathlib.

Reply:
Removed this confusing if statement, replacing it with faddr.with_suffix('.tiff'), which ensures the file will have the .tiff extension.